superset-frontend/src/explore/components/controls/AnnotationLayerControl/AnnotationLayer.jsx

```diff
@@ -298,74 +306,142 @@ class AnnotationLayer extends React.PureComponent {
     });
   }
 
-  fetchOptions(annotationType, sourceType, isLoadingOptions) {
-    if (isLoadingOptions) {
-      if (sourceType === ANNOTATION_SOURCE_TYPES.NATIVE) {
-        const queryParams = rison.encode({
-          page: 0,
-          page_size: FILTER_OPTIONS_LIMIT,
-        });
-        SupersetClient.get({
-          endpoint: `/api/v1/annotation_layer/?q=${queryParams}`,
-        }).then(({ json }) => {
-          const layers = json
-            ? json.result.map(layer => ({
-                value: layer.id,
-                label: layer.name,
-              }))
-            : [];
-          this.setState({
-            isLoadingOptions: false,
-            valueOptions: layers,
-          });
-        });
-      } else if (requiresQuery(sourceType)) {
-        const queryParams = rison.encode({
-          filters: [
-            {
-              col: 'id',
-              opr: 'chart_owned_created_favored_by_me',
-              value: true,
+  handleTextValue(inputValue) {
+    this.setState({
+      value: inputValue,
+    });
+  }
+
+  asyncFetch = async (_, page, pageSize) => {
+    const { annotationType, sourceType } = this.state;
+
+    if (sourceType === ANNOTATION_SOURCE_TYPES.NATIVE) {
+      const queryParams = rison.encode({
+        page,
+        page_size: pageSize,
+      });
+
+      const { json } = await SupersetClient.get({
+        endpoint: `/api/v1/annotation_layer/?q=${queryParams}`,
+      });
+
+      const { result, count } = json;
+
+      const layersObj = result.reduce((acc, layer) => {
+        acc[layer.id] = {
+          value: layer.id,
+          label: layer.name,
+        };
+        return acc;
+      }, {});
+
+      const layersArray = Object.values(layersObj);
+
+      this.setState(prevState => ({
```

Comment:
You shouldn't have a `setState` inside the async function. To fix this, instead of caching all the options loaded by the select and augmenting this state with chart data, you can have a specific state entry to store that information. You can use the annotation layer id as the key and chart data as the value.

Reply:
Could you elaborate on this one a little more? Do you mean get rid of the `valueOptions` state and save only the selected single chart in a state entry?

Reply:
Yes. You don't need to store the options in the state, they are already maintained by the Select component. You need the chart data information associated with a particular option. You can have a state entry specifically for that like a map of option id and chart data to support multiple charts considering that users might change selected options while interacting with the component.

Reply:
@rtexelm I'll try to explain my comment in more depth:

If you check the `asyncFetch` function, you'll see that there's a call to the chart endpoint:

```
const { json } = await SupersetClient.get({
   endpoint: `/api/v1/chart/?q=${queryParams}`,
});
```

If you check what's actually being used from the result you'll see `chart.id`, `chart.slice_name`, `chart.form_data` and `chart.viz_type`. So the first step is to change `queryParams` to only select these columns, using the `columns` key, to avoid fetching more data than needed.

You need the slice information when selecting an option to use in `renderSliceConfiguration`. Here you have 2 options:

1 - Instead of loading `chart.form_data` for all charts in `asyncFetch`, you could load only the `id`, `slice_name` and `viz_type`, and fire another query to load the `form_data` when a user selects a chart, keeping the previous loaded charts in case a user changes the selection. That's what I meant by:

> You can have a state entry specifically for that like a map of option id and chart data to support multiple charts considering that users might change selected options while interacting with the component.

2 - If you want to fire only one query, loading all `form_data` at once, even if probably only one will be used, you don't need a separate state to store that. The `onChange` event of the Select component sends the `value` as the first parameter and the `option` as the second parameter. Your `option` already contains the slice:

```
{
   value: chart.id,
   label: chart.slice_name,
   slice: {...}
}
```

Reply:
@michael-s-molina I'm not sure if this newest revision is close to what you were thinking but I can always continue to refine it